### PR TITLE
Internal updates

### PR DESCRIPTION
**Pull request by nakirekommula**
### Description
Experimental changes.

***
### Testing
None.

***

### Type of Change
Place an `x` the applicable box:
- [ ] Bug fix. Add the issue # below if applicable.
- [x] New feature. A non-breaking change which adds functionality.
- [ ] Other, such as a build process or documentation change.
***

#### Notes
None.

### CODE REVIEW

File: gma/src/ios/GADNativeAdCpp.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+@interface GADNativeAd ()
+
+/// Used only by allowlisted ad units. Provide a dictionary containing click data.
+- (void)performClickWithData:(nonnull NSDictionary *)clickData;
+
+/// Used only by allowlisted ad units. Provide a dictionary containing impression data. Returns YES
+/// if the impression is successfully recorded, otherwise returns NO.
+- (BOOL)recordImpressionWithData:(nonnull NSDictionary *)impressionData;
+
+@end
```
[the remark this answers is not in the document]
Comment (nakirekommula) [reply gma/src/ios/GADNativeAdCpp.h]: Done.

File: gma/src/ios/GADNativeAdCpp.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+#import <GoogleMobileAds/GADNativeAd.h>
+#import <UIKit/UIKit.h>
+
+@interface GADNativeAd ()
```
[the remark this answers is not in the document]
Comment (nakirekommula) [reply gma/src/ios/GADNativeAdCpp.h]: Ack.

**Review by berile — APPROVED** [review]
LGTM! Some minor suggestions. Thank you!

File: gma/integration_test/src/integration_test.cc [hunk 3]
```diff
@@ -2058,6 +2058,61 @@ TEST_F(FirebaseGmaTest, TestNativeAdLoadEmptyRequest) {
   delete native_ad;
 }
 
+TEST_F(FirebaseGmaTest, TestNativeRecordImpressionAndClick) {
```
[the remark this answers is not in the document]
Comment (nakirekommula) [reply gma/integration_test/src/integration_test.cc]: Ack.

File: gma/src/common/native_ad.cc [hunk 4]
```diff
@@ -84,5 +85,35 @@ const std::vector<NativeAdImage>& NativeAd::images() const {
   return internal_->images();
 }
 
+Future<void> NativeAd::RecordImpression(const Variant& impression_data) {
+  if (!impression_data.is_map()) {
```
[the remark this answers is not in the document]
Comment (nakirekommula) [reply gma/src/common/native_ad.cc:89]: API counterparts on Android/iOS do accept empty bundle/Dictionary. So, it should be okay to accept empty map here.

File: gma/src/common/native_ad.cc [hunk 5]
```diff
@@ -84,5 +85,35 @@ const std::vector<NativeAdImage>& NativeAd::images() const {
...
+  return internal_->RecordImpression(impression_data);
+}
+
+Future<void> NativeAd::RecordImpressionLastResult() const {
+  return internal_->GetLastResult(
+      firebase::gma::internal::kNativeAdFnRecordImpression);
+}
+
+Future<void> NativeAd::PerformClick(const Variant& click_data) {
+  if (!click_data.is_map()) {
```
[the remark this answers is not in the document]
Comment (nakirekommula) [reply gma/src/common/native_ad.cc:104]: API counterparts on Android/iOS do accept empty bundle/Dictionary. So, it should be okay to accept empty map here.

**Comment by nakirekommula** [comment]
> LGTM! Some minor suggestions. Thank you!

Thanks for the review! Addressed the comments in a new commit.

File: gma/src/ios/native_ad_internal_ios.mm [hunk 6]
```diff
@@ -64,6 +69,103 @@
...
+      [dict setObject:val_dict
+                     forKey:@(key.string_value())];
+    } else {
+      // Unsupported value type.
+      return nullptr;
+    }
+  }
+
+  NSDictionary *ret = [dict copy];
+  return ret;
```
Comment (jonsimantov) [inline gma/src/ios/native_ad_internal_ios.mm:166]: Why does this need to copy at the end?

**Review by jonsimantov — APPROVED** [review]
Looks good, just a few nits/questions.

File: gma/src/ios/native_ad_internal_ios.mm (hunk 6 above)
Comment (nakirekommula) [reply gma/src/ios/native_ad_internal_ios.mm:166]: API accepts NSDictionary as input. We initially use an NSMutableDictionary while generating the input, and finally convert it to an NSDictionary using the fully populated NSMutableDictionary.

File: gma/src/ios/native_ad_internal_ios.mm [hunk 7]
```diff
@@ -64,6 +69,103 @@
...
+
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [(GADNativeAd*)native_ad_ performClickWithData:click_payload];
+    CompleteFuture(kAdErrorCodeNone, nullptr, future_handle, &future_data_);
+  });
+
+  return future;
+}
+
+NSDictionary* NativeAdInternalIOS::variantmap_to_nsdictionary(const Variant &variant_data) {
```
[the remark this answers is not in the document]
Comment (nakirekommula) [reply gma/src/ios/native_ad_internal_ios.mm:131]: This is an internal helper method and so tried following similar lower case naming as other internal helper methods  ex: is_initialized

File: gma/src/android/native_ad_internal_android.cc [hunk 8]
```diff
@@ -138,6 +138,131 @@ Future<AdResult> NativeAdInternalAndroid::LoadAd(const char* ad_unit_id,
...
+  env->CallVoidMethod(
+      helper_, native_ad_helper::GetMethodId(native_ad_helper::kPerformClick),
+      reinterpret_cast<jlong>(callback_data), click_bundle);
+
+  util::CheckAndClearJniExceptions(env);
+
+  return future;
+}
+
+jobject NativeAdInternalAndroid::variantmap_to_bundle(
```
[the remark this answers is not in the document]
Comment (nakirekommula) [reply gma/src/android/native_ad_internal_android.cc:209]: Same reason as mentioned in other comment.

**Comment by nakirekommula** [comment]
> Looks good, just a few nits/questions.

Thanks for the review Jon! Added responses to the comments. PTAL and let me know.

File: gma/src/android/native_ad_internal_android.cc [hunk 9]
```diff
@@ -138,6 +138,131 @@ Future<AdResult> NativeAdInternalAndroid::LoadAd(const char* ad_unit_id,
...
+                          key_str, val_str);
+      env->DeleteLocalRef(val_str);
+    } else if (value.is_map()) {
+      jobject val_bundle = variantmap_to_bundle(value);
+      env->CallVoidMethod(variant_bundle,
+                          util::bundle::GetMethodId(util::bundle::kPutBundle),
+                          key_str, val_bundle);
+      env->DeleteLocalRef(val_bundle);
+    } else {
+      // Unsupported value type.
```
Comment (DellaBitta) [inline gma/src/android/native_ad_internal_android.cc:255]: You need to remove the ref to the key_str here, too.

**Review by DellaBitta — CHANGES_REQUESTED** [review]
There's a rare code path that leaves a hanging jobject reference.

File: gma/src/android/native_ad_internal_android.cc (hunk 9 above)
Comment (nakirekommula) [reply gma/src/android/native_ad_internal_android.cc:255]: Good catch! Addressed in the latest commit.

**Comment by nakirekommula** [comment]
> There's a rare code path that leaves a hanging jobject reference.

Thanks for the review David! Addressed the comments in the latest commit.

File: gma/src/android/native_ad_internal_android.cc [hunk 10]
```diff
@@ -138,6 +138,133 @@ Future<AdResult> NativeAdInternalAndroid::LoadAd(const char* ad_unit_id,
...
+
+  for (const auto& kvp : variant_data.map()) {
+    const Variant& key = kvp.first;
+    const Variant& value = kvp.second;
+
+    if (!key.is_string()) {
+      return nullptr;
+    }
+    jstring key_str = env->NewStringUTF(key.string_value());
+
```
Comment (DellaBitta) [inline gma/src/android/native_ad_internal_android.cc:231]: Might want to add `util::CheckAndClearJniExceptions(env);` here.

File: gma/src/android/native_ad_internal_android.cc [hunk 11]
```diff
@@ -138,6 +138,133 @@ Future<AdResult> NativeAdInternalAndroid::LoadAd(const char* ad_unit_id,
...
+      env->DeleteLocalRef(val_str);
+    } else if (value.is_map()) {
+      jobject val_bundle = variantmap_to_bundle(value);
+      env->CallVoidMethod(variant_bundle,
+                          util::bundle::GetMethodId(util::bundle::kPutBundle),
+                          key_str, val_bundle);
+      env->DeleteLocalRef(val_bundle);
+    } else {
+      // Unsupported value type.
+      util::CheckAndClearJniExceptions(env);
```
Comment (DellaBitta) [inline gma/src/android/native_ad_internal_android.cc:256]: you can remove `util::CheckAndClearJniExceptions(env);` here.

**Review by DellaBitta — APPROVED** [review]
Approved with some last nit comments!

File: gma/src/android/native_ad_internal_android.cc (hunk 10 above)
Comment (nakirekommula) [reply gma/src/android/native_ad_internal_android.cc:232]: Done.

File: gma/src/android/native_ad_internal_android.cc (hunk 11 above)
Comment (nakirekommula) [reply gma/src/android/native_ad_internal_android.cc]: Done.

**Comment by nakirekommula** [comment]
> Approved with some last nit comments!

Thanks again!

**Pull request merged**